modules/javafx.base/src/main/java/com/sun/javafx/binding/ListenerListBase.java

```diff
@@ -0,0 +1,456 @@
+                CHANGE_LISTENERS.set(this, i, null);
+
+                containsNulls = true;
+            }
+        }
+
+        if (containsNulls) {
+
+            /*
+             * Note: only nulls are removed here. Expired weak listeners are not removed
+             * as this would mean the listener count could change unexpectedly at the end
+             * of a notification without going through an addListener/removeListener call
+             * which may be overridden to track listeners. A scenario that would be troubling
+             * is:
+             *
+             * - A notification starts
+             * - An unrelated listener removes itself while list is locked (which will trigger this
+             *   clean-up code later); this goes through the normal add/removeListener channel
+             * - The notification ends
+             * - While unlocking, a weak listener is determined to be expired
+             *   - Removing this weak listener would not go through proper channels and thus
+             *     code that overrides add/removeListener would be unaware of the change
+             *
+             * Therefore, weak listeners will only be actively removed when another listener
+             * is being added or removed as the caller will then be expecting a change in the
+             * listener count, albeit higher than just the one listener being added or removed.
+             * Callers already should be aware that removing a listener may not change the count
+             * (if the listener didn't exist), so manually checking the new count after such a
+             * call must be done already.
+             */
+
+            INVALIDATION_LISTENERS.removeIf(this, Objects::isNull);
+            CHANGE_LISTENERS.removeIf(this, Objects::isNull);
+
+            nulledListenerCount = 0;
+        }
+
+        lockedSize = -1;
+
+        return true;
+    }
+
+    /**
+     * Locks this list.
+     */
+    protected final void lock() {
+        assertNotLocked();
+
+        this.lockedSize = invalidationListenersCount + changeListenersCount;
+    }
+
+    /**
+     * Checks whether this list is locked.
+     *
+     * @return {@code true} if locked, otherwise {@code false}
+     */
+    protected final boolean isLocked() {
+        return lockedSize >= 0;
+    }
+
+    /**
+     * Checks whether this list has any change listeners. Note: this does not
+     * update while locked.
+     *
+     * @return {@code true} if there were change listeners, otherwise {@code false}
+     */
+    public final boolean hasChangeListeners() {
+        return totalSize() > invalidationListenersCount;
+    }
+
+    private int totalSize() {
+        return isLocked() ? lockedSize : invalidationListenersCount + changeListenersCount;
+    }
+
+    private void assertInvalidationListenerIndex(int index) {
+        assert index < invalidationListenersCount : index + " >= " + invalidationListenersCount + ", results would be undefined";
```

Comment:
Do we allow `assert` in production code? This will require to run the application with `ea`. Or are these intended to be skipped in production and are for tests only?

Reply:
They're useful during testing (when `-ea` is passed, which it generally is). They will help catch bugs if this code is ever refactored.  As for the current implementation, I'm confident that it will never trip these asserts.  It also serves as a bit of documentation.  I'm fine with removing these -- I placed them in separate methods to not mess up inlining of compiled code (asserts will count against the method length limit if placed directly in the method).  As it is now, they will be completely eliminated when not running with `-ea`.

As far as allowing goes -- FX is full of asserts already, so I guess its allowed?

Reply:
I think I left these in because of the need to subclass these, and the subclasses are expected to work in a certain way and must be highly optimized.  Normally, I'd just make them hard checks (with `IllegalStateException` or `IllegalArgumentException`), but for optimal performance it is better to do it with asserts in this case (I don't want it to be too much slower than `ExpressionHelper`).  With the asserts, if you write any kind of test code for your subclasses, then it will fail quickly if you made a mistake.  As said, I'm confident these subclasses that we have now are implemented correctly, but who knows, in the future we may want to use similar code for ListChangeListeners or something...